### PR TITLE
Fixed mozilla/thimble.mozilla.org#1887: File overwrites without notifying user

### PR DESCRIPTION
**Pull request by simon66**
Fixed bug where imported items could overwrite local files without notifying the user (As shown here: https://github.com/mozilla/thimble.mozilla.org/issues/1887).

While working on the file, I've added some spaces (for better code indentation). Let me know if thats something you want reverted 👍

### CODE REVIEW

**Comment by simon66** [comment]
By the way, I just noticed that this PR needs to land first

https://github.com/mozilla/brackets/pull/659

I've changed the title to WIP. Once that lands, I'll add my fix 👍

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 1]
```diff
@@ -95,42 +143,52 @@ define(function (require, exports, module) {
...
-                                if(err) {
-                                    return callback(err);
-                                }
-                                fs.writeFile(path.absPath, path.data, callback);
-                            });
-                        } else {
-                            fs.writeFile(path.absPath, path.data, callback);
-                        }
+                            if (stats.type === "FILE") {
+                                //console.log("File exists!", path.absPath);
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: Remove this please.

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 2]
```diff
@@ -215,54 +281,79 @@ define(function (require, exports, module) {
...
             var basedir = Path.dirname(path);
 
-            if(_skipFile(path)) {
+            if (_skipFile(path)) {
                 return callback();
             }
 
-            fs.mkdirp(basedir, function(err) {
-                if(err && err.code !== "EEXIST") {
+            fs.mkdirp(basedir, function (err) {
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: It looks to me like we're duplicating code here from above.  Is there any way to refactor this to a function that both call?

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 3]
```diff
@@ -215,54 +281,79 @@ define(function (require, exports, module) {
...
+                if (err) {
                     return callback(err);
                 }
 
-                fs.writeFile(path, new Buffer(data), {encoding: null}, callback);
+                fs.stat(path, function (err, stats) {
+                    if (err && err.code !== "ENOENT") {
+                        return callback(err);
+                    }
+                    if (stats.type === "FILE") {
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: If you don't care about any case other that `"FILE"` do this:

```js
if (stats.type !== "FILE") {
    return callback();
}
// continue without indenting
...
```

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 4]
```diff
@@ -16,31 +15,40 @@ define(function (require, exports, module) {
...
+    var OVERWRITE_OPERATION = 1;
+    var KEEP_EXISITING_OPERATION = 2;
 
     // Mac and Windows clutter zip files with extra files/folders we don't need
     function _skipFile(filename) {
         var basename = Path.basename(filename);
 
         // Skip OS X additions we don't care about in the browser fs
-        if(/^__MACOSX\//.test(filename)) {
+        if (/^__MACOSX\//.test(filename)) {
```
Comment (gideonthomas) [inline src/filesystem/impls/filer/ArchiveUtils.js]: undo all these space changes

File: src/filesystem/impls/filer/ArchiveUtils.js (hunk 2 above)
Comment (simon66) [inline src/filesystem/impls/filer/ArchiveUtils.js]: You're right. I'm going to put this code into a function and test it out.

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 5]
```diff
@@ -281,35 +280,37 @@ define(function (require, exports, module) {
...
 
-            if (_skipFile(path)) {
+            if(_skipFile(path)) {
                 return callback();
             }
 
-            fs.mkdirp(basedir, function (err) {
-                if (err) {
+            fs.mkdirp(basedir, function(err) {
+                if(err) {
```
Comment (simon66) [inline src/filesystem/impls/filer/ArchiveUtils.js]: You're right @humphd. I'm going to put this code into a function and test it out.

**Comment by humphd** [comment]
I've landed #659 so, feel free to rebase and finish this.

**Comment by simon66** [comment]
I just did a rebase but I'm having a hard time debugging the new code. I enabled "async" on my debugger but the break point is not being picked up.

This might take a bit to complete. If this is an urgent issue that needs to be completed soon, let me know. I'll continue working on it :)

**Comment by simon66** [comment]
Wasn't as hard as I thought.
For some odd reason, the import doesn't trigger if the .tar archive contains empty files (0 bytes). Not too sure how files are stored when they are tar'd.

Going to remove the `[WIP]` label.
Please let me know if there's anything you want changed :+1:

File: locales/en-US/editor.properties [hunk 6]
```diff
@@ -136,6 +139,7 @@ DND_SUCCESS_UNTAR_TITLE=Untar Completed Successfully
 DND_SUCCESS_UNZIP=Successfully unzipped <b>{0}</b>.
 # {0} will be replaced by a tar filename
 DND_SUCCESS_UNTAR=Successfully untarred <b>{0}</b>.
+DND_FILE_REPLACE=A file named \"{0}\" already exists in this location. Do you want to use the imported file or keep the existing?
```
Comment (humphd) [inline locales/en-US/editor.properties]: Have all these strings been added to Thimble's repo yet?

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 7]
```diff
@@ -16,6 +16,15 @@ define(function (require, exports, module) {
     var Buffer          = Filer.Buffer;
     var Path            = Filer.Path;
     var fs              = Filer.fs();
+    var Dialogs         = require("widgets/Dialogs");
+    var DefaultDialogs  = require("widgets/DefaultDialogs");
+    var Strings         = require("strings");
+    var StringUtils     = require("utils/StringUtils");
+
+    // These are const variables
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: Get rid of this comment.

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 8]
```diff
@@ -100,21 +152,30 @@ define(function (require, exports, module) {
...
-                            });
-                        } else {
-                            fs.writeFile(path.absPath, path.data, callback);
+                    fs.mkdirp(basedir, function (err) {
+                        if (err) {
+                            return callback(err);
                         }
+                        fs.stat(path.absPath, function(err, stats) {
+                            if(err && err.code !== "ENOENT") {
+                                    return callback(err);
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: Too much indent here.

File: src/filesystem/impls/filer/ArchiveUtils.js [hunk 9]
```diff
@@ -227,11 +288,33 @@ define(function (require, exports, module) {
             }
 
             fs.mkdirp(basedir, function(err) {
-                if(err && err.code !== "EEXIST") {
```
Comment (humphd) [inline src/filesystem/impls/filer/ArchiveUtils.js]: Why this change?

File: src/filesystem/impls/filer/lib/WebKitFileImport.js [hunk 10]
```diff
@@ -134,13 +135,13 @@ define(function (require, exports, module) {
...
 
-        function handleRegularFile(deferred, file, filename, buffer, encoding) {
+        function saveFile(deferred, file, filename, buffer, encoding) {
             file.write(buffer, {encoding: encoding}, function(err) {
                 if (err) {
                     onError(deferred, filename, err);
                     return;
                 }
-
+
```
Comment (humphd) [inline src/filesystem/impls/filer/lib/WebKitFileImport.js]: Get rid of this change.

File: src/filesystem/impls/filer/lib/WebKitFileImport.js [hunk 11]
```diff
@@ -150,11 +151,55 @@ define(function (require, exports, module) {
             });
         }
 
+        function handleRegularFile(deferred, file, filename, buffer, encoding) {
+            fs.exists(filename, function(doesExist) {
+                if (doesExist) {
+                    console.log("File: ", filename, " already exists!");
```
Comment (humphd) [inline src/filesystem/impls/filer/lib/WebKitFileImport.js]: Remove this.

File: src/filesystem/impls/filer/lib/WebKitFileImport.js [hunk 12]
```diff
@@ -150,11 +151,55 @@ define(function (require, exports, module) {
...
+                        ]
+                    )
+                    .done(function(id) {
+                        if (id === Dialogs.DIALOG_BTN_IMPORT) {
+                            // Override file per user's request
+                            saveFile(deferred, file, filename, buffer, encoding);
+                        }
+                    });
+                } else {
+                    // File doesn't exist. Save without prompt
```
Comment (humphd) [inline src/filesystem/impls/filer/lib/WebKitFileImport.js]: Do this case first, and then return, and then you don't need to indent everything above so much:

```js
if (!doesExist) {
    // File doesn't exist. Save without prompt
    saveFile(deferred, file, filename, buffer, encoding);
    return;
}

// rest of code here for other case.
```

File: src/utils/DragAndDrop.js [hunk 13]
```diff
@@ -182,6 +181,7 @@ define(function (require, exports, module) {
             return result.promise();
         }, false)
             .fail(function () {
+                console.log("fail");
```
Comment (humphd) [inline src/utils/DragAndDrop.js]: Revert all the changes in this file.

**Comment by gideonthomas** [comment]
@simon66 are you still working on this?

**Comment by humphd** [comment]
If he's not, I'll finish it.

**Comment by humphd** [comment]
Finishing this in https://github.com/mozilla/brackets/pull/847
